PowerPoint_Adapter/AdapterActions/Push.cs

```diff
@@ -119,67 +91,72 @@ public override List<object> Push(IEnumerable<object> objects, string tag = "",
                 try
                 {
                     if (m_OutputFileSettings != null)
-                        presentationDoc.Clone(m_OutputFileSettings.GetFullFileName());
+                        presentationDoc.Clone(m_OutputFileSettings.GetFullFileName()).Dispose(); // Dispose the cloned document to avoid a memory leak
                     else if (m_OutputStream != null)
                     {
                         presentationDoc.Clone(m_OutputStream);
                         m_OutputStream.Position = 0;
                     }
-
-                }
-                catch (Exception e)
-                {
-                    BH.Engine.Base.Compute.RecordError("Could not save the changes: " + e.Message);
-                }
-
-            }
-            finally
-            {
-                // Release all content from memory
-                if (presentationDoc != null)
-                {
-                    presentationDoc.Dispose();
                 }
-                if (memoryStream != null)
+                catch (Exception ex)
                 {
-                    memoryStream.Close();
+                    BH.Engine.Base.Compute.RecordError(ex, "An error occurred while trying to save the changes");
+                    return new List<object>();
                 }
             }
 
-
             return objects.ToList();
         }
 
         /***************************************************/
         /**** Private Methods                           ****/
         /***************************************************/
 
-        private MemoryStream OpenTemplateFile(string filePath)
+        private PresentationDocument OpenTemplateDocument()
         {
-            // Make sure the file exists
-            if (!File.Exists(filePath))
+            MemoryStream memoryStream = new MemoryStream();
+
+            if (m_TemplateFileSettings != null)
+                memoryStream = OpenTemplateFile(m_TemplateFileSettings.GetFullFileName());
+            else if (m_TemplateStream != null)
+                m_TemplateStream.CopyTo(memoryStream);
+            else
+            {
+                BH.Engine.Base.Compute.RecordError("Neither a template file settings nor a template file stream could be found to get the template file from.");
+                return null;
+            }
+
+            try
+            {
+                return PresentationDocument.Open(memoryStream, true);
+            }
+            catch (Exception ex)
             {
-                BH.Engine.Base.Compute.RecordError($"There is no presentation with the file path {filePath}");
+                BH.Engine.Base.Compute.RecordError(ex, "An error occurred while trying to open the template presentation.");
                 return null;
             }
```

Comment:
- Creation of the stream directly in line 117 instead of initializing it as null means that an empty memory stream will be replaced in line 120 without being disposed
 - The catch doesn't dispose of the memory stream. This is only fine if the presentation has already registered the stream in its resources  before the exception is thrown and if the presentation disposes of the stream when being disposed. Can we be 100% sure those two ifs are always true ?